File: src/main/java/mnm/mods/tabbychat/net/SNetworkVersion.java
package mnm.mods.tabbychat.net;

import mnm.mods.tabbychat.TabbyChat;
import mnm.mods.tabbychat.api.Channel;
import mnm.mods.tabbychat.client.ChatManager;
import net.minecraft.network.PacketBuffer;
import net.minecraft.util.text.TranslationTextComponent;
import net.minecraftforge.fml.network.NetworkEvent;

import java.util.function.Supplier;

public class SNetworkVersion {

    private String version;

    public SNetworkVersion(PacketBuffer buffer) {
        version = buffer.readString(20);
    }

    public SNetworkVersion(String version) {
        this.version = version;
    }

    public void encode(PacketBuffer buffer) {
        buffer.writeString(version);
    }

    public void handle(Supplier<NetworkEvent.Context> context) {
        context.get().enqueueWork(() -> {
            ChatManager chat = ChatManager.instance();
            Channel chan = chat.getChannel("TabbyChat");

            if (!version.equals(TabbyChat.PROTOCOL_VERSION)) {
                chat.addMessage(chan, new TranslationTextComponent("tabbychat.network.mismatch", TabbyChat.PROTOCOL_VERSION, version));
            }
        });
        context.get().setPacketHandled(true);
    }
}

File: src/main/kotlin/mnm/mods/tabbychat/net/SNetworkVersion.kt
package mnm.mods.tabbychat.net

import mnm.mods.tabbychat.PROTOCOL_VERSION
import mnm.mods.tabbychat.client.gui.NotificationToast
import mnm.mods.tabbychat.util.mc
import net.minecraft.network.PacketBuffer
import net.minecraft.util.text.TranslationTextComponent
import net.minecraftforge.fml.network.NetworkEvent
import java.util.function.Supplier

class SNetworkVersion {

    val version: String

    constructor(buffer: PacketBuffer) {
        version = buffer.readString(20)
    }

    constructor(version: String) {
        this.version = version
    }

    fun encode(buffer: PacketBuffer) {
        buffer.writeString(version)
    }

    fun handle(context: Supplier<NetworkEvent.Context>) {
        context.get().enqueueWork {

            if (version != PROTOCOL_VERSION) {
                val title = TranslationTextComponent("tabbychat.network.mismatch")
                mc.toastGui.add(NotificationToast("TabbyChat", title))
            }
        }
        context.get().packetHandled = true
    }
}
